Postcondition function with return value

feliperodri opened the pull request:
Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

*Description of changes:*

Define a function contract to encapsulate the check of a postcondition with the return value of a function. The cost in prod is 0, because when the `DEBUG` mode is turn off the condition is not even evaluated.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

feliperodri committed:
Postcondition function with return value

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

danielsn commented on include/aws/common/common.h:
```diff
@@ -41,6 +41,19 @@
+#define AWS_RETURN_SUCCESS_WITH_POSTCONDITION(cond) (AWS_OP_SUCCESS)
+#define AWS_RETURN_SIZE_WITH_POSTCONDITION(cond, size) (size)
+#else
+#define AWS_RETURN_SUCCESS_WITH_POSTCONDITION(cond) aws_assert_postcondition(cond)
```
Cleaner way
```
/* Use the comma operator to evaluate the assert, then return the value
 * https://en.wikipedia.org/wiki/Comma_operator
 */
#define AWS_RETURN_VALUE_WITH_POSTCONDITION(val, cond) (assert(cond), (val))
#define AWS_RETURN_SUCCESS_WITH_POSTCONDITION(cond)  AWS_RETURN_VALUE_WITH_POSTCONDITION(AWS_OP_SUCCESS, (cond))
```

feliperodri replied:
great, I'm on it!

danielsn commented on include/aws/common/common.h:
```diff
@@ -41,6 +41,20 @@
+ * DEBUG mode is turn off the condition is not even evaluated. This keep the
+ * assert in the same line for debuggability.
+ */
+#ifdef NDEBUG
```
You don't need the ndebug check anymore, since the assert will already handle that.

danielsn commented on include/aws/common/common.h:
```diff
@@ -50,20 +50,24 @@
+ * If NDEBUG is on, this will have 0 runtime cost.
+ */
+#define AWS_RETURN_VALUE_WITH_POSTCONDITION(cond, val) (assert(cond), (val))
+#define AWS_RETURN_SUCCESS_WITH_POSTCONDITION(cond) AWS_RETURN_VALUE_WITH_POSTCONDITION((cond), AWS_OP_SUCCESS)
```
What macro should we use for void functions?

feliperodri replied:
`AWS_POSTCONDITION` only? As we currently have

tautschnig commented on include/aws/common/array_list.inl:
```diff
@@ -80,9 +80,8 @@ bool aws_array_list_is_valid(const struct aws_array_list *AWS_RESTRICT list) {
     size_t required_size;
     bool required_size_is_valid = aws_mul_size_checked(list->length, list->item_size, &required_size) == AWS_OP_SUCCESS;
     bool current_size_is_valid = list->current_size >= required_size;
-    bool item_size_is_valid = list->item_size != 0;
```
Unrelated change?

tautschnig reviewed (commented):
I'm worried about the (seemingly) unrelated change. Otherwise it seems ok, but not super convincing to me: I personally find the existing code easier to read.

bretambrose committed:
Date time work (#310)


* Add basic iso 8601 parsing and output

feliperodri committed:
Refactor aws_array_list harnesses with new proof style (#313)

Refactored functions:
 - aws_array_list_capacity
 - aws_array_list_clean_up
 - aws_array_list_clear
 - aws_array_list_copy

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

markrtuttle commented on .cbmc-batch/jobs/aws_array_list_capacity/aws_array_list_capacity_harness.c:
```diff
@@ -16,36 +16,28 @@
-
 /**
- * Runtime: 0m4.951s
+ * Runtime: 5s
```
Still not sure this data belongs in the source code, seems like a maintenance headache.

feliperodri replied:
@danielsn @tautschnig any thoughts?

markrtuttle commented on .cbmc-batch/jobs/aws_array_list_back/cbmc-batch.yaml:
```diff
@@ -1,4 +1,4 @@
 jobos: ubuntu16
-cbmcflags: "--bounds-check;--pointer-check;--div-by-zero-check;--signed-overflow-check;--unsigned-overflow-check;--pointer-overflow-check;--undefined-shift-check;--float-overflow-check;--nan-check;--unwinding-assertions;--unwindset;memcpy_impl.0:3;--function;aws_array_list_back_harness"
+cbmcflags: "--bounds-check;--pointer-check;--div-by-zero-check;--signed-overflow-check;--unsigned-overflow-check;--pointer-overflow-check;--undefined-shift-check;--float-overflow-check;--nan-check;--unwinding-assertions;--unwindset;memcpy_impl.0:3;--unwind;1;--object-bits;8"
```
Object bits 8 bits is the default.

markrtuttle commented on .cbmc-batch/jobs/aws_array_list_capacity/aws_array_list_capacity_harness.c:
```diff
@@ -16,36 +16,28 @@
+    /* save current state of the data structure */
+    struct aws_array_list old = list;
+    struct store_byte_from_buffer old_byte;
+    save_byte_from_array((uint8_t *)list.data, list.current_size, &old_byte);
```
This is a trick we use to prove no byte in the array has changed.  I don't think this will be clear to someone outside the proof community.  This is one thing that might be worth explaining in a top-level readme file somewhere.

markrtuttle commented on .cbmc-batch/jobs/aws_array_list_clean_up/aws_array_list_clean_up_harness.c:
```diff
@@ -16,30 +16,26 @@
+    assert(list.current_size == 0);
+    assert(list.length == 0);
+    assert(list.item_size == 0);
+    assert(list.data == NULL);
```
I feel surprised to be reading predicates with fairly descriptive names, and then to find this "null" predicate spelled out in the code in a style different from the others.  Not a big deal.

markrtuttle commented on .cbmc-batch/jobs/aws_array_list_clear/aws_array_list_clear_harness.c:
```diff
@@ -16,35 +16,29 @@
+    assert(list.current_size == old.current_size);
+    assert(list.item_size == old.item_size);
+    assert(list.data == old.data);
 }
```
I don't know the semantics of "clear", but this is a surprising list of assertions.  In particular, if there is list data, then the list length 0, but if there is no data then the list lengths are unchanged.

markrtuttle commented on .cbmc-batch/jobs/aws_array_list_copy/aws_array_list_copy_harness.c:
```diff
@@ -16,61 +16,30 @@
+    assert(aws_array_list_is_valid(&from));
+    assert(aws_array_list_is_valid(&to));
+    assert(from.item_size == to.item_size);
 }
```
No check of the data copied?

markrtuttle commented on source/byte_buf.c:
```diff
@@ -414,9 +414,7 @@ int aws_byte_buf_append_with_lookup(
-    AWS_POSTCONDITION(aws_byte_cursor_is_valid(from));
-    return AWS_OP_SUCCESS;
+    return AWS_RETURN_SUCCESS_WITH_POSTCONDITION(aws_byte_cursor_is_valid(from) && aws_byte_buf_is_valid(to));
 }
```
Is this macro really easier to read that the original code?  It is done, maybe not worth changing, but I would reconsider this style in future modules.

markrtuttle commented on tests/CMakeLists.txt:
```diff
@@ -260,8 +260,11 @@ add_test_case(rfc822_invalid_format)
 add_test_case(rfc822_invalid_auto_format)
 add_test_case(iso8601_utc_parsing)
+add_test_case(iso8601_basic_utc_parsing)
 add_test_case(iso8601_utc_parsing_auto_detect)
```
The existence of the source/date_time.c code is pretty mysterious to me in a proof package.  It would probably help if there were documentation in the file that you are writing tests.
